ConfigMap Watcher got nil Event

# What's going on?

Injector deployment crash frequently. I got crash log:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xef548b]

goroutine 35 [running]:
github.com/tumblr/k8s-sidecar-injector/internal/pkg/config/watcher.(*K8sConfigMapWatcher).Watch(0xc0002cca00, 0x12d1f60, 0xc0002d8600, 0xc000300600, 0x0, 0x0)
	/src/internal/pkg/config/watcher/watcher.go:109 +0x36b
main.main.func1.1(0xc0002cca00, 0x12d1f60, 0xc0002d8600, 0xc000300600)
	/src/cmd/main.go:114 +0x77
created by main.main.func1
	/src/cmd/main.go:111 +0xd7
```
 I think it just because watcher got a Event with a `nil Object` and try to use it. I find this bug is caused by https://github.com/kubernetes/client-go/issues/334

# Expected Behavior

# Reproducer

can reproduce with any valid sidecar configurations, just wait a few minutes

# Version Deets

* Kubernetes Version: `v1.13.3`
* `k8s-sidecar-injector` Version: `release-v0.1.6`

--- a/cmd/main.go
+++ b/cmd/main.go
@@ -113,7 +113,12 @@ func main() {
 				glog.Infof("launching watcher for ConfigMaps")
 				err := configWatcher.Watch(ctx, sigChan)
 				if err != nil {
-					glog.Fatalf("error watching for new ConfigMaps (terminating): %s", err.Error())
+					switch err {
+					case watcher.EventNilObjectError:
+						glog.Errorf("watcher got error, try to restart watcher: %s", err.Error())
+					default:
+						glog.Fatalf("error watching for new ConfigMaps (terminating): %s", err.Error())
+					}
 				}
 			}
 		}()

--- a/cmd/main.go
+++ b/cmd/main.go
@@ -114,7 +114,7 @@ func main() {
 				err := configWatcher.Watch(ctx, sigChan)
 				if err != nil {
 					switch err {
-					case watcher.EventNilObjectError:
+					case watcher.WatchChannelClosedError:
 						glog.Errorf("watcher got error, try to restart watcher: %s", err.Error())
 					default:
 						glog.Fatalf("error watching for new ConfigMaps (terminating): %s", err.Error())

--- a/internal/pkg/config/watcher/watcher.go
+++ b/internal/pkg/config/watcher/watcher.go
@@ -12,6 +12,7 @@ import (
 	"github.com/golang/glog"
 	"github.com/tumblr/k8s-sidecar-injector/internal/pkg/config"
 	"k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/watch"
@@ -115,6 +116,9 @@ func (c *K8sConfigMapWatcher) Watch(ctx context.Context, notifyMe chan<- interfa
 				glog.Errorf("channel has closed, should restart watcher")
 				return WatchChannelClosedError
 			}
+			if e.Type == watch.Error {
+				return apierrs.FromObject(e.Object)
+			}
 			glog.V(3).Infof("event: %s %s", e.Type, e.Object.GetObjectKind())
 			switch e.Type {
 			case watch.Added:
@@ -132,7 +136,6 @@ func (c *K8sConfigMapWatcher) Watch(ctx context.Context, notifyMe chan<- interfa
 		case <-ctx.Done():
 			glog.V(2).Infof("stopping configmap watcher, context indicated we are done")
 			// clean up, we cancelled the context, so stop the watch
-			watcher.Stop()
 			return nil
 		}
 	}

--- a/internal/pkg/config/watcher/watcher.go
+++ b/internal/pkg/config/watcher/watcher.go
@@ -4,6 +4,7 @@ package watcher
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"strings"
@@ -23,6 +24,9 @@ import (
 const (
 	serviceAccountNamespaceFilePath = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
 )
+
+// EventNilObjectError: should restart watcher
+var EventNilObjectError = errors.New("got watcher event with nil Object")
 
 // K8sConfigMapWatcher is a struct that connects to the API and collects, parses, and emits sidecar configurations
 type K8sConfigMapWatcher struct {
@@ -106,10 +110,12 @@ func (c *K8sConfigMapWatcher) Watch(ctx context.Context, notifyMe chan<- interfa
 	for {
 		select {
 		case e = <-ch:
+			// after receive event with nil Object, watcher will begin receive thousands of events with nil Object
+			// restart watcher can fix this bug
+			// detail at https://github.com/kubernetes/client-go/issues/334
 			if e.Object == nil {
-				glog.Errorf("got nil event, restart watcher")
-				watcher.Stop()
-				return nil
+				glog.Errorf("got event with nil Object, should restart watcher")
+				return EventNilObjectError
 			}
 			glog.V(3).Infof("event: %s %s", e.Type, e.Object.GetObjectKind())
 			switch e.Type {

--- a/internal/pkg/config/watcher/watcher.go
+++ b/internal/pkg/config/watcher/watcher.go
@@ -25,8 +25,8 @@ const (
 	serviceAccountNamespaceFilePath = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
 )
 
-// EventNilObjectError: should restart watcher
-var EventNilObjectError = errors.New("got watcher event with nil Object")
+// WatchChannelClosedError: should restart watcher
+var WatchChannelClosedError = errors.New("watcher channel has closed")
 
 // K8sConfigMapWatcher is a struct that connects to the API and collects, parses, and emits sidecar configurations
 type K8sConfigMapWatcher struct {
@@ -102,20 +102,18 @@ func (c *K8sConfigMapWatcher) Watch(ctx context.Context, notifyMe chan<- interfa
 	watcher, err := c.client.ConfigMaps(c.Namespace).Watch(metav1.ListOptions{
 		LabelSelector: mapStringStringToLabelSelector(c.ConfigMapLabels),
 	})
+	defer watcher.Stop()
 	if err != nil {
 		return fmt.Errorf("unable to create watcher (possible serviceaccount RBAC/ACL failure?): %s", err.Error())
 	}
-	ch := watcher.ResultChan()
-	var e watch.Event
 	for {
 		select {
-		case e = <-ch:
-			// after receive event with nil Object, watcher will begin receive thousands of events with nil Object
-			// restart watcher can fix this bug
+		case e, ok := <-watcher.ResultChan():
+			// channel may closed caused by HTTP timeout, should restart watcher
 			// detail at https://github.com/kubernetes/client-go/issues/334
-			if e.Object == nil {
-				glog.Errorf("got event with nil Object, should restart watcher")
-				return EventNilObjectError
+			if !ok {
+				glog.Errorf("channel has closed, should restart watcher")
+				return WatchChannelClosedError
 			}
 			glog.V(3).Infof("event: %s %s", e.Type, e.Object.GetObjectKind())
 			switch e.Type {

--- a/internal/pkg/config/watcher/watcher.go
+++ b/internal/pkg/config/watcher/watcher.go
@@ -106,6 +106,11 @@ func (c *K8sConfigMapWatcher) Watch(ctx context.Context, notifyMe chan<- interfa
 	for {
 		select {
 		case e = <-ch:
+			if e.Object == nil {
+				glog.Errorf("got nil event, restart watcher")
+				watcher.Stop()
+				return nil
+			}
 			glog.V(3).Infof("event: %s %s", e.Type, e.Object.GetObjectKind())
 			switch e.Type {
 			case watch.Added:

--- a/internal/pkg/config/watcher/watcher_test.go
+++ b/internal/pkg/config/watcher/watcher_test.go
@@ -1,6 +1,9 @@
 package watcher
 
 import (
+	"context"
+	"k8s.io/apimachinery/pkg/watch"
+	testcore "k8s.io/client-go/testing"
 	"testing"
 
 	_ "github.com/tumblr/k8s-sidecar-injector/internal/pkg/testing"
@@ -28,5 +31,25 @@ func TestGet(t *testing.T) {
 	}
 	if len(messages) != 0 {
 		t.Fatalf("expected 0 messages, but got %d", len(messages))
+	}
+}
+
+func TestWatcherChannelClose(t *testing.T) {
+	client := fake.NewSimpleClientset()
+	watcher := watch.NewEmptyWatch()
+	client.PrependWatchReactor("configmaps", testcore.DefaultWatchReactor(watcher, nil))
+
+	w := K8sConfigMapWatcher{
+		Config: testConfig,
+		client: client.CoreV1(),
+	}
+
+	sigChan := make(chan interface{}, 10)
+	// background context never canceled, no deadline
+	ctx := context.Background()
+
+	err := w.Watch(ctx, sigChan)
+	if err != nil && err != WatchChannelClosedError {
+		t.Errorf("expect catch WatchChannelClosedError, but got %s", err)
 	}
 }

